File: python/hql/hql/store.py
"""Read-only HedronDB store. Opens sqlite via URI mode=ro; never writes."""

from __future__ import annotations

import sqlite3
from functools import lru_cache
from pathlib import Path
from typing import Optional
from urllib.parse import quote

from hql.row import Row, parse_extra_map

# Crate-root schema.sql is the only schema source (copy of vault store.sql).
# The Python twin reads that file; it never carries its own column list.
SCHEMA_SQL_PATH = Path(__file__).resolve().parents[3] / "schema.sql"

_CONSTRAINT_KEYWORDS = {"UNIQUE", "PRIMARY", "FOREIGN", "CHECK", "CONSTRAINT"}


def schema_sql() -> str:
    """Contents of crate-root schema.sql (executed by Store and by fixtures)."""
    return SCHEMA_SQL_PATH.read_text(encoding="utf-8")


def _split_top_level(body: str) -> list[str]:
    """Split on commas outside parentheses so UNIQUE (a, b) stays one piece."""
    pieces: list[str] = []
    depth = 0
    start = 0
    for i, ch in enumerate(body):
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth = max(depth - 1, 0)
        elif ch == "," and depth == 0:
            pieces.append(body[start:i])
            start = i + 1
    pieces.append(body[start:])
    return pieces


def parse_schema_sql(sql: str) -> dict[str, list[tuple[str, str]]]:
    """Table name -> (column, declared type) pairs in declaration order.

    Table constraints (UNIQUE, PRIMARY KEY, ...) are not columns and are
    skipped; `--` comments are stripped first. Mirrors the Rust reader.
    """
    stripped = "\n".join(line.split("--", 1)[0] for line in sql.splitlines())
    tables: dict[str, list[tuple[str, str]]] = {}
    rest = stripped
    while True:
        start = rest.find("CREATE TABLE")
        if start < 0:
            break
        after = rest[start + len("CREATE TABLE") :]
        open_ = after.find("(")
        close = after.find(");")
        if open_ < 0 or close < 0:
            break
        name = after[:open_].split()[-1]
        cols: list[tuple[str, str]] = []
        for piece in _split_top_level(after[open_ + 1 : close]):
            words = piece.split()
            if not words or words[0].upper() in _CONSTRAINT_KEYWORDS:
                continue
            cols.append((words[0], words[1] if len(words) > 1 else ""))
        tables[name] = cols
        rest = after[close + 2 :]
    return tables


@lru_cache(maxsize=1)
def expected_schema() -> dict[str, list[tuple[str, str]]]:
    return parse_schema_sql(schema_sql())


class Store:
    """Read-only sqlite3 handle. Never migrate; report schema drift only."""

    def __init__(self, db: str | Path) -> None:
        path = Path(db).expanduser().resolve()
        if not path.exists():
            raise FileNotFoundError(f"store not found: {path}")
        # sqlite URI: file:/abs/path?mode=ro — never CREATE, never write.
        uri = f"file:{quote(path.as_posix(), safe='/')}?mode=ro"
        self.path = path
        self.conn = sqlite3.connect(uri, uri=True)
        self.conn.row_factory = sqlite3.Row

    def schema_mismatches(self) -> list[str]:
        existing = self._table_columns()
        mismatches: list[str] = []
        expected = expected_schema()
        for table, cols in expected.items():
            if table not in existing:
                mismatches.append(f"missing table {table}")
                continue
            got = existing[table]
            want_names = [name for name, _ in cols]
            got_names = [name for name, _ in got]
            if got_names != want_names:
                mismatches.append(
                    f"table {table} columns {got_names} != {want_names}"
                )
            for (want_name, want_type), (got_name, got_type) in zip(cols, got):
                if want_name != got_name:
                    continue
                if want_type.upper() != got_type.upper():
                    mismatches.append(
                        f"table {table} column {want_name} type {got_type!r} != {want_type!r}"
                    )
        for table in existing:
            if table.startswith("sqlite_"):
                continue
            if table not in expected:
                mismatches.append(f"unexpected table {table}")
        return mismatches

    def nodes(self) -> list[Row]:
        rows = self.conn.execute(
            "SELECT id, vault_id, type, path, extra FROM nodes"
        ).fetchall()
        return [
            Row(
                path=row["path"],
                extra=row["extra"] or "",
                extra_map=parse_extra_map(row["extra"]),
                node_id=row["id"],
                vault_id=row["vault_id"],
                node_type=row["type"],
            )
            for row in rows
        ]

    def edges(self) -> list[dict[str, Optional[str]]]:
        rows = self.conn.execute(
            "SELECT id, vault_id, from_id, to_id, to_raw, type, properties FROM edges"
        ).fetchall()
        return [
            {
                "id": row["id"],
                "vault_id": row["vault_id"],
                "from_id": row["from_id"],
                "to_id": row["to_id"],
                "to_raw": row["to_raw"],
                "type": row["type"],
                "properties": row["properties"] or "",
            }
            for row in rows
        ]

    def latest_desired_states(self) -> dict[str, dict]:
        """Warm path: newest desired_states row per vault_id. Never reads events."""
        rows = self.conn.execute(
            "SELECT id, vault_id, state_version, reconciled_by, importance, spec, status "
            "FROM desired_states"
        ).fetchall()
        latest: dict[str, dict] = {}
        for row in rows:
            vault_id = row["vault_id"]
            version = int(row["state_version"])
            prev = latest.get(vault_id)
            if prev is not None and version <= prev["state_version"]:
                continue
            latest[vault_id] = {
                "id": row["id"],
                "vault_id": vault_id,
                "state_version": version,
                "reconciled_by": row["reconciled_by"],
                "importance": row["importance"],
                "spec": row["spec"] or "",
                "status": row["status"] or "",
            }
        return latest

    def causal_chain(self, desired_state_id: str) -> list[dict]:
        """Cool path: events that reconcile this Desired State. No spec/status/data."""
        rows = self.conn.execute(
            "SELECT id, vault_id, ts, actor, type, caused_by, reconciles, supersedes "
            "FROM events "
            "WHERE reconciles = ? "
            "ORDER BY ts ASC, id ASC",
            (desired_state_id,),
        ).fetchall()
        return [
            {
                "id": row["id"],
                "vault_id": row["vault_id"],
                "ts": int(row["ts"]),
                "actor": row["actor"],
                "type": row["type"],
                "caused_by": _format_caused_by(row["caused_by"]),
                "reconciles": row["reconciles"],
                "supersedes": row["supersedes"],
            }
            for row in rows
        ]

    def vault_ids_named(self, name: str) -> list[str]:
        ids: list[str] = []
        for row in self.conn.execute(
            "SELECT id, path, extra FROM nodes WHERE type = 'Vault'"
        ).fetchall():
            extra_map = parse_extra_map(row["extra"])
            if row["path"] == name or extra_map.get("name") == name:
                ids.append(row["id"])
        return ids

    def _table_columns(self) -> dict[str, list[tuple[str, str]]]:
        tables = [
            row[0]
            for row in self.conn.execute(
                "SELECT name FROM sqlite_master WHERE type = 'table'"
            ).fetchall()
        ]
        out: dict[str, list[tuple[str, str]]] = {}
        for table in tables:
            cols = [
                (row[1], row[2] or "")
                for row in self.conn.execute(f"PRAGMA table_info({table})").fetchall()
            ]
            out[table] = cols
        return out

    def close(self) -> None:
        self.conn.close()


def _format_caused_by(raw: Optional[str]) -> Optional[str]:
    text = (raw or "").strip()
    if not text or text in ("[]", "null", "~"):
        return None
    ids: list[str] = []
    for line in text.splitlines():
        line = line.strip().strip(",")
        if line.startswith("- "):
            ids.append(line[2:].strip().strip("'").strip('"'))
        elif line.startswith("[") and line.endswith("]"):
            inner = line[1:-1].strip()
            if inner:
                ids.extend(
                    part.strip().strip("'").strip('"')
                    for part in inner.split(",")
                    if part.strip()
                )
    return ",".join(ids) if ids else None

File: src/hql/store.rs
//! Read-only HedronDB store. rusqlite OPEN_READ_ONLY; never writes.

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::OnceLock;

use rusqlite::{Connection, OpenFlags};

use crate::error::{Error, Result};
use crate::hql::row::{parse_extra_map, DesiredStateView, HistoryEventView, Row};
use crate::store::{Store, SCHEMA_SQL};
use crate::types::Event;
use uuid::Uuid;

/// Expected tables and `(column, declared type)` pairs, derived from the
/// crate `schema.sql` embedded at compile time. Not a second handwritten list.
fn expected_schema() -> &'static [(String, Vec<(String, String)>)] {
    static PARSED: OnceLock<Vec<(String, Vec<(String, String)>)>> = OnceLock::new();
    PARSED.get_or_init(|| parse_schema_sql(SCHEMA_SQL))
}

/// Minimal `CREATE TABLE` reader for `schema.sql`: table name plus columns in
/// declaration order. Table constraints (`UNIQUE`, `PRIMARY KEY`, ...) are not
/// columns and are skipped. `--` comments are stripped first.
pub(crate) fn parse_schema_sql(sql: &str) -> Vec<(String, Vec<(String, String)>)> {
    const CONSTRAINTS: [&str; 5] = ["UNIQUE", "PRIMARY", "FOREIGN", "CHECK", "CONSTRAINT"];
    let stripped: String = sql
        .lines()
        .map(|line| line.split("--").next().unwrap_or(""))
        .collect::<Vec<_>>()
        .join("\n");
    let mut tables = Vec::new();
    let mut rest = stripped.as_str();
    while let Some(start) = rest.find("CREATE TABLE") {
        let after = &rest[start + "CREATE TABLE".len()..];
        let (Some(open), Some(close)) = (after.find('('), after.find(");")) else {
            break;
        };
        let name = after[..open].split_whitespace().last().unwrap_or("").to_string();
        let mut cols = Vec::new();
        for piece in split_top_level(&after[open + 1..close]) {
            let mut words = piece.split_whitespace();
            let Some(col) = words.next() else { continue };
            if CONSTRAINTS.contains(&col.to_ascii_uppercase().as_str()) {
                continue;
            }
            let ty = words.next().unwrap_or("").to_string();
            cols.push((col.to_string(), ty));
        }
        tables.push((name, cols));
        rest = &after[close + 2..];
    }
    tables
}

/// Split a column body on commas outside parentheses, so
/// `UNIQUE (vault_id, name)` stays one piece.
fn split_top_level(body: &str) -> Vec<&str> {
    let mut pieces = Vec::new();
    let mut depth = 0usize;
    let mut start = 0;
    for (i, c) in body.char_indices() {
        match c {
            '(' => depth += 1,
            ')' => depth = depth.saturating_sub(1),
            ',' if depth == 0 => {
                pieces.push(&body[start..i]);
                start = i + 1;
            }
            _ => {}
        }
    }
    pieces.push(&body[start..]);
    pieces
}

#[derive(Clone, Debug)]
pub(crate) struct EdgeRec {
    pub id: String,
    pub from_id: String,
    pub to_id: Option<String>,
    pub to_raw: Option<String>,
    pub edge_type: String,
    pub properties: String,
}

/// Read-only sqlite handle. Never migrate; report schema drift only.
pub struct RoStore {
    conn: Connection,
    path: PathBuf,
}

impl RoStore {
    pub fn open(db: impl AsRef<Path>) -> Result<Self> {
        let path = db.as_ref().to_path_buf();
        if !path.exists() {
            return Err(Error::Io(std::io::Error::new(
                std::io::ErrorKind::NotFound,
                format!("store not found: {}", path.display()),
            )));
        }
        let conn = Connection::open_with_flags(&path, OpenFlags::SQLITE_OPEN_READ_ONLY)?;
        Ok(Self { conn, path })
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    pub fn schema_mismatches(&self) -> Result<Vec<String>> {
        let existing = self.table_columns()?;
        let mut mismatches = Vec::new();
        let expected = expected_schema();
        let expected_names: Vec<&str> = expected.iter().map(|(n, _)| n.as_str()).collect();
        for (table, cols) in expected {
            let Some(got) = existing.get(table.as_str()) else {
                mismatches.push(format!("missing table {table}"));
                continue;
            };
            let want_names: Vec<&str> = cols.iter().map(|(n, _)| n.as_str()).collect();
            let got_names: Vec<&str> = got.iter().map(|(n, _)| n.as_str()).collect();
            if got_names != want_names {
                mismatches.push(format!(
                    "table {table} columns {got_names:?} != {want_names:?}"
                ));
            }
            for ((want_name, want_type), (got_name, got_type)) in cols.iter().zip(got.iter()) {
                if want_name != got_name {
                    continue;
                }
                if !want_type.eq_ignore_ascii_case(got_type) {
                    mismatches.push(format!(
                        "table {table} column {want_name} type {got_type:?} != {want_type:?}"
                    ));
                }
            }
        }
        for table in existing.keys() {
            if table.starts_with("sqlite_") {
                continue;
            }
            if !expected_names.contains(&table.as_str()) {
                mismatches.push(format!("unexpected table {table}"));
            }
        }
        Ok(mismatches)
    }

    pub fn nodes(&self) -> Result<Vec<Row>> {
        let mut stmt = self
            .conn
            .prepare("SELECT id, vault_id, type, path, extra FROM nodes")?;
        let rows = stmt.query_map([], |row| {
            let extra: Option<String> = row.get(4)?;
            let extra = extra.unwrap_or_default();
            let extra_map = parse_extra_map(Some(&extra));
            Ok(Row {
                path: row.get(3)?,
                extra,
                extra_map,
                node_id: row.get(0)?,
                vault_id: row.get(1)?,
                node_type: row.get(2)?,
                ..Row::default()
            })
        })?;
        rows.collect::<std::result::Result<Vec<_>, _>>()
            .map_err(Error::from)
    }

    pub(crate) fn edges(&self) -> Result<Vec<EdgeRec>> {
        let mut stmt = self
            .conn
            .prepare("SELECT id, vault_id, from_id, to_id, to_raw, type, properties FROM edges")?;
        let rows = stmt.query_map([], |row| {
            let properties: Option<String> = row.get(6)?;
            Ok(EdgeRec {
                id: row.get(0)?,
                from_id: row.get::<_, Option<String>>(2)?.unwrap_or_default(),
                to_id: row.get(3)?,
                to_raw: row.get(4)?,
                edge_type: row.get(5)?,
                properties: properties.unwrap_or_default(),
            })
        })?;
        rows.collect::<std::result::Result<Vec<_>, _>>()
            .map_err(Error::from)
    }

    /// Warm path: newest desired_states row per vault_id. Never reads events.
    pub fn latest_desired_states(&self) -> Result<HashMap<String, DesiredStateView>> {
        let mut stmt = self.conn.prepare(
            "SELECT id, vault_id, state_version, reconciled_by, importance, spec, status \
             FROM desired_states",
        )?;
        let rows = stmt.query_map([], |row| {
            let spec: Option<String> = row.get(5)?;
            let status: Option<String> = row.get(6)?;
            Ok(DesiredStateView {
                id: row.get(0)?,
                vault_id: row.get(1)?,
                state_version: row.get(2)?,
                reconciled_by: row.get(3)?,
                importance: row.get(4)?,
                spec: spec.unwrap_or_default(),
                status: status.unwrap_or_default(),
            })
        })?;
        let mut latest: HashMap<String, DesiredStateView> = HashMap::new();
        for row in rows {
            let row = row?;
            if let Some(prev) = latest.get(&row.vault_id) {
                if row.state_version <= prev.state_version {
                    continue;
                }
            }
            latest.insert(row.vault_id.clone(), row);
        }
        Ok(latest)
    }

    /// Cool path: `Store::causal_chain` event rows for one Desired State.
    /// Drops `data`; does not return spec vs status.
    pub fn causal_chain(&self, desired_state_id: &str) -> Result<Vec<HistoryEventView>> {
        let id =
            Uuid::parse_str(desired_state_id).map_err(|err| Error::Invalid(err.to_string()))?;
        let events = Store::read_causal_chain(&self.conn, id)?;
        Ok(events.iter().map(history_view).collect())
    }

    pub fn vault_ids_named(&self, name: &str) -> Result<Vec<String>> {
        let mut stmt = self
            .conn
            .prepare("SELECT id, path, extra FROM nodes WHERE type = 'Vault'")?;
        let rows = stmt.query_map([], |row| {
            let extra: Option<String> = row.get(2)?;
            Ok((
                row.get::<_, String>(0)?,
                row.get::<_, Option<String>>(1)?,
                extra.unwrap_or_default(),
            ))
        })?;
        let mut ids = Vec::new();
        for row in rows {
            let (id, path, extra) = row?;
            let extra_map = parse_extra_map(Some(&extra));
            let extra_name = extra_map.get("name").and_then(|v| v.as_deref());
            if path.as_deref() == Some(name) || extra_name == Some(name) {
                ids.push(id);
            }
        }
        Ok(ids)
    }

    fn table_columns(&self) -> Result<HashMap<String, Vec<(String, String)>>> {
        let mut tables = Vec::new();
        let mut stmt = self
            .conn
            .prepare("SELECT name FROM sqlite_master WHERE type = 'table'")?;
        let names = stmt.query_map([], |row| row.get::<_, String>(0))?;
        for name in names {
            tables.push(name?);
        }
        let mut out = HashMap::new();
        for table in tables {
            if !table.chars().all(|c| c.is_ascii_alphanumeric() || c == '_') {
                return Err(Error::Invalid(format!("unexpected table name {table}")));
            }
            let mut cols = Vec::new();
            let mut stmt = self.conn.prepare(&format!("PRAGMA table_info({table})"))?;
            let rows = stmt.query_map([], |row| {
                let name: String = row.get(1)?;
                let ty: String = row.get::<_, Option<String>>(2)?.unwrap_or_default();
                Ok((name, ty))
            })?;
            for row in rows {
                cols.push(row?);
            }
            out.insert(table, cols);
        }
        Ok(out)
    }
}

fn history_view(event: &Event) -> HistoryEventView {
    let caused_by = if event.caused_by.is_empty() {
        None
    } else {
        Some(
            event
                .caused_by
                .iter()
                .map(ToString::to_string)
                .collect::<Vec<_>>()
                .join(","),
        )
    };
    HistoryEventView {
        id: event.id.to_string(),
        vault_id: event.vault_id.to_string(),
        ts: event.ts,
        actor: event.actor.to_string(),
        event_type: event.event_type.clone(),
        caused_by,
        reconciles: event.reconciles.map(|id| id.to_string()),
        supersedes: event.supersedes.clone(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The parsed expectation must equal what sqlite reports after executing
    /// the same `schema.sql`; otherwise the drift check is checking a fiction.
    #[test]
    fn parsed_schema_matches_executed_schema() {
        let conn = Connection::open_in_memory().unwrap();
        conn.execute_batch(SCHEMA_SQL).unwrap();
        let parsed = parse_schema_sql(SCHEMA_SQL);
        assert!(!parsed.is_empty(), "schema.sql parsed to zero tables");
        for (table, cols) in &parsed {
            let mut stmt = conn.prepare(&format!("PRAGMA table_info({table})")).unwrap();
            let live: Vec<(String, String)> = stmt
                .query_map([], |row| Ok((row.get(1)?, row.get(2)?)))
                .unwrap()
                .collect::<std::result::Result<_, _>>()
                .unwrap();
            assert_eq!(&live, cols, "table {table}: PRAGMA vs parsed schema.sql");
        }
        let mut stmt = conn
            .prepare("SELECT name FROM sqlite_master WHERE type = 'table' AND name NOT LIKE 'sqlite_%' ORDER BY rowid")
            .unwrap();
        let live_tables: Vec<String> = stmt
            .query_map([], |row| row.get(0))
            .unwrap()
            .collect::<std::result::Result<_, _>>()
            .unwrap();
        let parsed_tables: Vec<String> = parsed.iter().map(|(n, _)| n.clone()).collect();
        assert_eq!(live_tables, parsed_tables);
    }
}
